rust_src/src/marker.rs

```diff
@@ -341,12 +337,121 @@ pub extern "C" fn marker_byte_position(marker: Lisp_Object) -> ptrdiff_t {
+    let b = LispObject::from_raw(buffer).as_buffer_or_current_buffer();
     if b.is_live() {
-        b.as_mut()
+        Some(b)
+    } else {
+        None
+    }
+}
+
+/// Internal function to set MARKER in BUFFER at POSITION.  Non-zero
+/// RESTRICTED means limit the POSITION by the visible part of BUFFER.
+fn set_marker_internal(
+    marker: LispObject,
+    position: LispObject,
+    buffer: LispObject,
+    restricted: bool,
+) -> LispObject {
+    let buf = live_buffer(buffer.to_raw());
+    let mut m = marker.as_marker_or_error();
+
+    // Set MARKER to point nowhere if BUFFER is dead, or
+    // POSITION is nil or a marker points to nowhere.
+    if position.is_nil()
+        || position
+            .as_marker()
+            .map_or(false, |m| !m.buffer().is_some()) || !buf.is_some()
+    {
+        unsafe {
+            unchain_marker(m.as_mut());
+        }
+
+    // Optimize the special case where we are copying the position of
+    // an existing marker, and MARKER is already in the same buffer.
+    } else if position.as_marker().map_or(false, |p| p.buffer() == buf) && m.buffer() == buf {
+        let pos = position.as_marker_or_error();
+        m.bytepos = pos.bytepos_or_error();
+        m.charpos = pos.charpos_or_error();
+    } else {
+        let b = buf.unwrap_or_else(|| panic!("Invalid buffer reference."));
+        set_marker_internal_else(m, position, restricted, b);
+    }
+    marker
+}
+
+fn set_marker_internal_else(
+    mut marker: LispMarkerRef,
+    position: LispObject,
+    restricted: bool,
+    mut buf: LispBufferRef,
+) {
+    let mut charpos: ptrdiff_t;
+    let mut bytepos: ptrdiff_t;
+
+    // Do not use CHECK_NUMBER_COERCE_MARKER because we
+    // don't want to call buf_charpos_to_bytepos if POSITION
+    // is a marker and so we know the bytepos already.
+    if let Some(num) = position.as_fixnum() {
+        charpos = num as ptrdiff_t;
+        bytepos = -1;
+    } else if let Some(m) = position.as_marker() {
+        charpos = m.charpos_or_error();
+        bytepos = m.bytepos_or_error();
+    } else {
+        wrong_type!(Qinteger_or_marker_p, position)
+    }
+    let beg = buffer_beg(buf, restricted);
+    let end = buffer_end(buf, restricted);
+    charpos = clip_to_bounds(beg, charpos as EmacsInt, end);
+
+    // Don't believe BYTEPOS if it comes from a different buffer,
+    // since that buffer might have a very different correspondence
+    // between character and byte positions.
+    if bytepos == -1
+        || !position
+            .as_marker()
+            .map_or(false, |m| m.buffer() == Some(buf))
+    {
+        bytepos = unsafe { buf_charpos_to_bytepos(buf.as_mut(), charpos) };
+    } else {
+        let beg = buffer_beg_byte(buf, restricted);
+        let end = buffer_end_byte(buf, restricted);
+        bytepos = clip_to_bounds(beg, bytepos as EmacsInt, end);
+    }
+    attach_marker(marker.as_mut(), buf.as_mut(), charpos, bytepos);
+}
+
```

Comment:
Would the following functions work better as methods on `LispBuffer`?

Reply:
I thought it would be bad to only pass booleans to a function/method ?

Reply:
The boolean now is no clearer. `(self, bool)` or `(LispBufferRef, bool)` are basically the same.

One option would be to pass an Enum instead. `BUFFER_VISIBLE` or `BUFFER_NOT_VISIBLE` maybe?

Reply:
I've decided to use a bool because this way I don't have to check `restricted`. Or do I miss something ?